### ISSUE TITLE
features used that should be behind #ifdef

### ISSUE THREAD

**Issue by igalic** [truenas/py-libzfs#20]
I've been hacking on a rather crude port of this library for Linux, and noticed that certain features are not properly guarded by conditionals, despite the fact that we test for them!

here's my patch:

```diff
diff --git a/configure.ac b/configure.ac
index ce0a71b..5b048c8 100644
--- a/configure.ac
+++ b/configure.ac
@@ -14,27 +14,35 @@ define(SECTION_TITLE,
 dnl works for checking define statements in c
 m4_include([m4/AX_CHECK_DEFINE.m4])
 
-freebsd_src=${FREEBSD_SRC:=/usr/src}
-
-dnl Setting appropriate CFLAGS
-
-CFLAGS="-nostdinc 
--I${freebsd_src}/cddl/lib/libumem 
--I${freebsd_src}/sys/cddl/compat/opensolaris/ 
--I${freebsd_src}/sys/cddl/compat/opensolaris 
--I${freebsd_src}/cddl/compat/opensolaris/include 
--I${freebsd_src}/cddl/compat/opensolaris/lib/libumem 
--I${freebsd_src}/cddl/contrib/opensolaris/lib/libzpool/common 
--I${freebsd_src}/sys/cddl/contrib/opensolaris/common/zfs 
--I${freebsd_src}/sys/cddl/contrib/opensolaris/uts/common/fs/zfs 
--I${freebsd_src}/sys/cddl/contrib/opensolaris/uts/common/sys 
--I${freebsd_src}/cddl/contrib/opensolaris/head 
--I${freebsd_src}/sys/cddl/contrib/opensolaris/uts/common 
--I${freebsd_src}/cddl/contrib/opensolaris/lib/libnvpair 
--I${freebsd_src}/cddl/contrib/opensolaris/lib/libuutil/common 
--I${freebsd_src}/cddl/contrib/opensolaris/lib/libzfs/common 
--I${freebsd_src}/cddl/contrib/opensolaris/lib/libzfs_core/common 
-${CFLAGS} -I/usr/include"
+if test `uname -s` == "Linux" ; then
+  ZFS_VERSION='0.7.5'
+  CFLAGS="
+  -I/usr/include/libspl
+  -I/usr/include/libzfs
+  "
+else
+  freebsd_src=${FREEBSD_SRC:=/usr/src}
+
+  dnl Setting appropriate CFLAGS
+
+  CFLAGS="-nostdinc 
+  -I${freebsd_src}/cddl/lib/libumem 
+  -I${freebsd_src}/sys/cddl/compat/opensolaris/ 
+  -I${freebsd_src}/sys/cddl/compat/opensolaris 
+  -I${freebsd_src}/cddl/compat/opensolaris/include 
+  -I${freebsd_src}/cddl/compat/opensolaris/lib/libumem 
+  -I${freebsd_src}/cddl/contrib/opensolaris/lib/libzpool/common 
+  -I${freebsd_src}/sys/cddl/contrib/opensolaris/common/zfs 
+  -I${freebsd_src}/sys/cddl/contrib/opensolaris/uts/common/fs/zfs 
+  -I${freebsd_src}/sys/cddl/contrib/opensolaris/uts/common/sys 
+  -I${freebsd_src}/cddl/contrib/opensolaris/head 
+  -I${freebsd_src}/sys/cddl/contrib/opensolaris/uts/common 
+  -I${freebsd_src}/cddl/contrib/opensolaris/lib/libnvpair 
+  -I${freebsd_src}/cddl/contrib/opensolaris/lib/libuutil/common 
+  -I${freebsd_src}/cddl/contrib/opensolaris/lib/libzfs/common 
+  -I${freebsd_src}/cddl/contrib/opensolaris/lib/libzfs_core/common 
+  ${CFLAGS} -I/usr/include"
+fi
 
 dnl Setting appropriate LDFLAGS
 LDFLAGS="-lzfs -lnvpair -lzfs_core -luutil -lgeom -v"
@@ -93,12 +101,21 @@ AC_COMPILE_IFELSE([AC_LANG_PROGRAM(
 SECTION_TITLE([RUNNING TESTS ON libzfs.h HEADER])
 
 
-AC_CHECK_HEADERS([libzfs.h], [], [AC_ERROR([A working libzfs header is required])], [
-	#include <sys/types.h>
-	#include <sys/mnttab.h>
-	#include <sys/fs/zfs.h>
-	#include <sys/zfs_ioctl.h>
-])
+if test `uname -s` == "Linux" ; then
+  AC_CHECK_HEADERS([libzfs.h], [], [AC_ERROR([A working libzfs header is required])], [
+    #include <sys/types.h>
+    #include <sys/mnttab.h>
+    #include <sys/fs/zfs.h>
+  ])
+else
+  AC_CHECK_HEADERS([libzfs.h], [], [AC_ERROR([A working libzfs header is required])], [
+    #include <sys/types.h>
+    #include <sys/mnttab.h>
+    #include <sys/fs/zfs.h>
+    #include <sys/zfs_ioctl.h>
+  ])
+fi
+
 
 
 dnl Checking function signature of zpool_scan
```

and here's the output from `./configure`:

```
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed

-----------------------------------
   RUNNING TESTS ON zfs.h HEADER
-----------------------------------
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... no
checking for sys/types.h... yes
checking for sys/stat.h... no
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for sys/types.h... (cached) yes
checking for sys/fs/zfs.h... yes
checking for ZFS_MAX_DATASET_NAME_LEN defined in sys/fs/zfs.h... yes
configure: Checking existence of pss_pass_scrub_pause & pss_pass_scrub_spent_paused struct member
configure: OK
configure: Checking typedef pool_scrub_cmt_t in zfs header
configure: OK

--------------------------------------
   RUNNING TESTS ON libzfs.h HEADER
--------------------------------------
checking for libzfs.h... yes
configure: Checking function signature of zpool_scan
configure: 2 params
configure: Checking function signature of zfs_send_one
configure: 4 params
configure: Checking EZFS_SCRUB_PAUSED presence
configure: OK
configure: Checking typedef sendflags_t in libzfs header
configure: OK
configure: Checking LZC_SEND_FLAG_EMBED_DATA presence
configure: OK
configure: Checking params for zfs_receive
configure: 6 params
configure: Checking existence of zfs_send_resume and zfs_send_resume_token_to_nvlist
configure: NOT FOUND
checking for ZFS_MAXNAMELEN defined in libzfs.h... no
checking for ZPOOL_MAXNAMELEN defined in libzfs.h... no
configure: Checking for resumable member in recvflags_t struct
configure: OK
configure: Checking for compress member in sendflags_t struct
configure: OK

-------------------------------------------
   RUNNING TESTS ON libzfs_core.h HEADER
-------------------------------------------
checking for libzfs_core.h... yes
configure: Verifying lzc_bookmark existence
configure: NOT FOUND
configure: lzc_send_space function not found

-------------------------
   GENERATING CONFIG.H
-------------------------
configure: creating ./config.status
config.status: creating pxd/config.pxi
config.status: creating Makefile
```

and here's the output from `make`:

```
/home/igalic/.pyenv/shims/python setup.py build
WARNING: '' not a valid package name; please use only .-separated package names in setup.py
running build
running build_py
running build_ext
cythoning libzfs.pyx to libzfs.c
warning: ./pxd/nvpair.pxd:75:2: 'DATA_TYPE_UINT8_ARRAY' redeclared 

Error compiling Cython file:
------------------------------------------------------------
...



IF HAVE_LZC_SEND_FLAG_EMBED_DATA:
    class SendFlags(enum.IntEnum):
        EMBED_DATA = libzfs.LZC_SEND_FLAG_EMBED_DATA
                          ^
------------------------------------------------------------

libzfs.pyx:323:27: cimported module has no attribute 'LZC_SEND_FLAG_EMBED_DATA'

Error compiling Cython file:
------------------------------------------------------------
...

        with nogil:
            IF HAVE_LZC_SEND_SPACE == 4:
                ret = libzfs.lzc_send_space(c_name, cfromname, 0, &space)
            ELSE:
                ret = libzfs.lzc_send_space(c_name, cfromname, &space)
                           ^
------------------------------------------------------------

libzfs.pyx:2242:28: cimported module has no attribute 'lzc_send_space'

Error compiling Cython file:
------------------------------------------------------------
...
        bookmarks = NVList()
        bookmarks['{0}#{1}'.format(self.parent.name, name)] = self.name
        c_bookmarks = bookmarks.handle

        with nogil:
            ret = libzfs.lzc_bookmark(c_bookmarks, NULL)
                       ^
------------------------------------------------------------

libzfs.pyx:2606:24: cimported module has no attribute 'lzc_bookmark'

Error compiling Cython file:
------------------------------------------------------------
...

        with nogil:
            IF HAVE_LZC_SEND_SPACE == 4:
                ret = libzfs.lzc_send_space(c_name, cfromname, 0, &space)
            ELSE:
                ret = libzfs.lzc_send_space(c_name, cfromname, &space)
                                          ^
------------------------------------------------------------

libzfs.pyx:2242:43: Calling gil-requiring function not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
        bookmarks = NVList()
        bookmarks['{0}#{1}'.format(self.parent.name, name)] = self.name
        c_bookmarks = bookmarks.handle

        with nogil:
            ret = libzfs.lzc_bookmark(c_bookmarks, NULL)
                                    ^
------------------------------------------------------------

libzfs.pyx:2606:37: Calling gil-requiring function not allowed without gil
building 'libzfs' extension
creating build
creating build/temp.linux-x86_64-3.6
gcc…
libzfs.c:1:2: error: #error Do not use this file, it is the result of a failed Cython compilation.
 #error Do not use this file, it is the result of a failed Cython compilation.
  ^~~~~
error: command 'gcc' failed with exit status 1
Makefile:5: recipe for target 'build' failed
make: *** [build] Error 1
```

**Comment by sonicaj** [comment]
You changed the compiler flags for the autoconf script, did you do the same for py-libzfs setup.py ?

**Comment by igalic** [comment]
not enough, i'm afraid.
perhaps we should generate setup.py from configure…

**Comment by sonicaj** [comment]
That is something we can think about. I think it is safe to close this issue then as we have concluded the issue is different to what outlined in the description. Please if you have other suggestions, feel free to comment or open up a new issue. Ty

**Comment by igalic** [comment]
@sonicaj i'm confused.
the problem is not the last `error` from gcc, but the myriad of Cython errors before that.
all relating to bookmark and send feature, which have been correctly recognized by configure:

```
DEF HAVE_RECVFLAGS_T_RESUMABLE = 1
DEF HAVE_LZC_BOOKMARK = 0
DEF HAVE_ZFS_MAX_DATASET_NAME_LEN = 1
DEF HAVE_POOL_SCRUB_CMD_T = 1
DEF HAVE_LZC_SEND_SPACE = 0
DEF HAVE_ZPOOL_SCAN = 2
DEF HAVE_ZFS_SEND_ONE = 4
DEF HAVE_ZFS_RECEIVE = 6
DEF HAVE_ZFS_MAXNAMELEN = 0
DEF HAVE_ZPOOL_MAXNAMELEN = 0
DEF HAVE_EZFS_SCRUB_PAUSED = 1
DEF HAVE_ZFS_SEND_RESUME = 0
DEF HAVE_ZFS_SEND_RESUME_TOKEN_TO_NVLIST = 0
DEF HAVE_SENDFLAGS_T_COMPRESS = 1
DEF HAVE_LZC_SEND_FLAG_EMBED_DATA = 1
DEF HAVE_SENDFLAGS_T_TYPEDEF = 1
DEF HAVE_POOL_SCAN_STAT_T_PAUSE = 1
```

but are not respected by the source code.

**Comment by sonicaj** [comment]
I think I can help with that. The way cython currently searches for packages is that it is configured to do so thinking of freebsd only and sets the appropriate compiler flags. Now when we try to build it, of course the cython compiler is unable to find the necessary libzfs.package because the place it is looking at most probably doesn't exist. You'll need to make sure that when cython searches for the header files, it is looking in the correct place

**Comment by sonicaj** [comment]
Now configure and cython are two different things. Configure runs it's own flow with it's own compiler flags which is configured to look in specific places. If it is able to find those files, that doesn't mean that the cython project will do so as well. Cython project's compiler also needs to be configured to look in the right place. The only relation between them right now is the assumption that they have the same flags set and thus if configure says something is there, it is safe to believe in cython that it is. But in this case, we violate this relation as they don't have the same flags set. Hope this clears a few things

**Comment by akkornel** [comment]
Hello!  For what it's worth, I was able to make a little more progress on this.  All of those Cython errors stem from the configure script thinking that some of the `libzfs_core.h` stuff is missing (the "NOT FOUND" message in the `configure` output).  That was happening because `-lgeom` is in `LDFLAGS`, and Linux does not do GEOM.  Removing `-lgeom` and re-running `autoconf` allows you to make progress.

I was able to get a build to the point where I was able to identify differences between libzfs on FreeBSD and libzfs on Linux.  Unfortunately, I'm not able to continue my work right now, but I left all my nodes in #21.  I hope it helps!
